src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -130,11 +131,23 @@ private MessageCacheItem createCacheItem(Iterator<Locale> fallbackLocalesIter) {
 		CacheService cacheService = new CacheService(dto);
 		// Create a new cacheItem object to be stored in cache
 		MessageCacheItem cacheItem = new MessageCacheItem();
-		refreshCacheItem(cacheItem, VIPCfg.getInstance().getMsgOriginsQueue().iterator());
 
+		ProductService ps = new ProductService(dto);
+		if (!ps.isSupportedLocale(Locale.forLanguageTag(dto.getLocale()))) {
+			Locale matchedLocale = LocaleUtility.pickupLocaleFromList(new LinkedList<>(ps.getSupportedLocales()), Locale.forLanguageTag(dto.getLocale()));
+			if (ps.isSupportedLocale(matchedLocale)) {
+				MessagesDTO matchedLocaleDTO = new MessagesDTO(dto.getComponent(), matchedLocale.toLanguageTag(), dto.getProductID(), dto.getVersion());
+				cacheItem = new ComponentService(matchedLocaleDTO).getMessages(null);
+				MessageCacheItem cacheItemCopy = new MessageCacheItem(matchedLocale.toLanguageTag(), null, null, System.currentTimeMillis(), cacheItem.getMaxAgeMillis());
+				cacheService.addCacheOfComponent(cacheItemCopy);
+				return cacheItem;
+			}
+		}
```

Comment:
@huihuiw01 , you are missing some use cases in the above code.
1. We want fallbackLocalesIter to be a parameter of the method.  If you trace, getMessages eventually calls createCacheItem if cahceItem is not yet in cache. In there, if fallbackLocalesIter is null, no fallback is applied. We need this for use cases where fallback is not necessary.
2. A "matched locale" does not necessarily mean that it is supported. For example: supported locales may include "fr" and "en_US". If the requested locale is "fr_CA", it matches "fr", but fr_CA" is technically NOT supported. For this, the client will use "fr" bundle only because fr_CA bundle does not exist. If the service is updated to support fr_CA (i.e. supported locales are now "fr", "en_US" and "fr_CA"), then the client must not use the locale matching logic so that it will pick up "fr_CA" from the service.

That being said, you are right, we do not need to cache non-supported locales with an empty map anymore because we are now using getSupportedLocales in here. However, if you look at the implementation of ProductService.getSupportedLocales, it is not yet using the cache store this list. Hence, every call to it is expensive. So for now, the call to it is in createCacheItem only. I plan to update this in PR754 which is planned for next Sprint.

Reply:
@huihuiw01 , you are missing a lot of use cases in the above code.

1. We need to create a new CacheItem if it is not found in cache (line 119).
2. We need to check for expiry of MessageCacheItem objects and refresh the content accordingly. These are in lines 108-109, 116
3. We want fallbackLocalesIter to be a parameter of the method. If you trace, the getMessages method eventually calls createCacheItem (if cacheItem is not yet in cache). In there, if fallbackLocalesIter is null, no fallback is applied. We need this for use cases where fallback is not necessary.
4. A "matched locale" does not necessarily mean that it is "supported". For example: supported locales may include "fr" and "en_US". If the requested locale is "fr_CA", it matches "fr", but fr_CA" is technically NOT supported. For this, the client will use "fr" bundle only because "fr_CA" bundle does not exist. Later on, if the service is updated to support "fr_CA" (i.e. supported locales are now "fr", "en_US" and "fr_CA"), then the client must not use the locale matching logic so that it will pick up "fr_CA" from the service. That is why there is another "if" in line 138.
5. There is a reason why I am caching non-supported locales with an empty cacheData map instead of directly serving the next available fallback. We want the client to eventually check the service if the non-supported locale has been made available in the service. In other words, if the locale is now supported. The only way to do this is to add an "expiry" that is specific to the cached non-supported locale so that when it expires, it means it is time to check again from the VIP service. Note: We can not check from VIP service for every request due to performance impact so we wait for expiry of the non-supported locale. This is how it is implemented:

For a non-supported locale, a MessageCacheItem with an empty cacheData map, the current timestamp, and the maxAgeMillis of the fallback locale is created. This MessageCacheItem is added to cache using the non-supported locale in the cache key. This logic is in line 157-158.
Now when this MessageCacheItem is accessed, the cacheData map is empty and so the corresponding fallback locale's data map has to be served (lines 112 -113). 

I believe #5 is very much related to my cache expiry logic comments in your PR. Do you need to connect over a phone call to further clarify?
https://github.com/vmware/singleton/pull/742#discussion_r482185125
https://github.com/vmware/singleton/pull/742#discussion_r482210946

Reply:
@huihuiw01 , you are missing some use cases in your code suggestion above.

1. We need to fetch data from data store and create a new CacheItem if data is not found in cache (see line 119).
2. We need to check for expiry of MessageCacheItem objects and refresh the content accordingly. These are in lines 108-109, 116
3. We want fallbackLocalesIter to be a parameter of the method. If you trace getMessages, it eventually calls createCacheItem (if cacheItem is not yet in cache). In createCacheItem, if fallbackLocalesIter is null, no fallback is applied. We need this for use cases where fallback is not necessary.

That being said, yes we do not need to cache non-supported locales with an empty cacheData map anymore because we are now using ProductService.getSupportedLocales to identify which is supported and which is not. However, if you look at ProductService.getSupportedLocales, it is not yet using the cache, so each call to this method is expensive (as of now). Hence, I was planning to postpone code clean up to 757.  What do you think?